### PR TITLE
Loss

### PR DESCRIPTION
**Pull request by YoojLee**
## Loss 수정
#1 Loss가 정상적으로 산출되지 않는 문제 해결
-> 확인 결과, label이 positive pair끼리 묶는 방식으로 해결이 된 게 아니었음. 따라서, 이와 관련해서 label 만드는 부분 수정.

## SyncBatchNorm 수정
SyncBatchNorm 같은 경우 모든 프로세스에서 batch statistic을 계산하고, 이를 sync하도록 구현되어 있음. 따라서, validation을 하나의 process에서만 수행할 때에 syncbatchnorm이 다른 프로세스로부터 batch statistic을 전달받고자 무한 대기하는 현상이 발생함.
-> validation을 모든 프로세스에서 수행하는 방향으로 문제 해결.

## Optimizer 변경
하나의 optimizer에서 representation과 classifier의 가중치를 모두 업데이트하는 방식으로 수행한 결과, classifier에서도 learning rate scheduling이 일어나는 문제 발생.
-> 각 부분마다 별도의 optimizer를 둠으로써 해결. 다만, 이 경우 learning curve가 매우 완만해지는데 optimizer의 종류가 sgd여서 변형된 건지는 잘 모르겠음.

### CODE REVIEW

File: loss.py [hunk 1]
```diff
@@ -14,13 +14,9 @@ def forward(self, sim, y):
...
             sim: a similarity matrix (2N X 2N)
+            y: labels
         """
-        # masking a diagonal of a similarity matrix.
-        mask = torch.ones_like(sim)
-        mask.fill_diagonal_(LARGE_NUMBER) # large number 채우기
-
-        # apply a mask to a similarity matrix.
-        sim *= mask
+        sim.fill_diagonal_(LARGE_NUMBER)
```
Comment (YoojLee) [inline loss.py:19]: 이 부분 같은 경우에는 원래 masking을 곱해주는 형식이었는데, 생각해보니 LARGE NUMBER를 곱해준다고 해서 동일하게 LARGE NUMBER가 산출된다는 보장이 없음. 또한, 어차피 fill_diagonal을 활용할 것이라면 그냥 similarity에 바로 fill_diagonal을 활용하는 게 맞을 것임.

File: train.py [hunk 2]
```diff
@@ -162,25 +171,26 @@ def main_worker(local_rank, ngpus_per_node, opt):
...
         
         # forward
         output = model(image)
 
         sim = get_cosine_similarity(output)
         
+        # generate labels
         y = []
-        for i in range(sim.size(0)//2):
+        for i in range(0,sim.size(0),2): # 이거 이렇게 안하고 더 빨리할 수 있는 방법 없나..?
```
Comment (YoojLee) [inline train.py:189]: 이 부분은 for loop 안쓰고 torch 내부 메소드를 이용해서 변경해볼 수 있을까 하는데 일단 학습 정상화가 되면 리팩토링 하는 게 좋을 것 같다.

File: train.py [hunk 3]
```diff
@@ -89,20 +87,23 @@ def main_worker(local_rank, ngpus_per_node, opt):
 
         from torchlars import LARS
 
-        base_optimizer = torch.optim.SGD(
-            [{'params': model.parameters()},
-            {'params': classifier.parameters(), 'lr': opt.eval_lr}],
-             lr=opt.lr, weight_decay=opt.weight_decay)
+        base_optimizer_repr = torch.optim.SGD(
```
Comment (YoojLee) [inline train.py:90]: base optimizer 종류도 opt로 받아올 수 있도록 하는 게 좋을 것 같은데 이렇게 될 경우에는 optimizer 별로 다르게 하이퍼 파라미터가 적용되는 부분은 어떻게 반영할지 조금 더 알아봐야할 것 같음. 물론 json 파일로 configuration을 관리하면 괜찮을 것 같긴 한데..최대한 argument parser를 활용하고 싶으니까...

File: train.py [hunk 4]
```diff
@@ -203,99 +213,88 @@ def train(train_loader, model, classifier, criterion, criterion_c, optimizer, sc
...
+            image = image.view(-1, 3, opt.resize, opt.resize).to(opt.local_rank)
 
-            sim = get_cosine_similarity(model(image))
+            sim = get_cosine_similarity(model(image)) # sim: [2N, 2N]
             
             y = []
 
-            for i in range(sim.size(0)//2):
-                y.append(i+1) 
+            for i in range(0,sim.size(0),2): # N iterations
```
Comment (YoojLee) [inline train.py:265]: 이 부분도 마찬가지. 굳이 iteration 안 돌고 토치 내부에서 해결할 수 있는지 확인해볼 것.
